src/dataflow/src/source/mod.rs

```diff
@@ -1340,15 +1344,45 @@ where
                     )
                 })?;
 
-                for (pid, offset) in offsets {
-                    timestamp_histories.add_partition(pid, offset);
+                let starting_offsets = source_persist.get_starting_offsets(valid_bindings.iter());
+
+                tracing::debug!(
+                    "In {}, initial (restored) source offsets: {:?}. upper_bindings_seal_ts = {}, upper_data_seal_ts = {}",
+                    name,
+                    starting_offsets,
+                    source_persist.config.upper_bindings_seal_ts,
+                    source_persist.config.upper_data_seal_ts,
+                );
+
+                tracing::debug!(
+                    "In {}, initial (restored) timestamp bindings: valid_bindings: {:?}, latest_bindings: {:?}",
+                    name,
+                    valid_bindings, latest_bindings
+                );
+
+                for (pid, offset) in starting_offsets {
+                    if crate::source::responsible_for(
```

Comment:
this (and the subsequent uses of `responsible_for` feel like "restore returned all of the bindings, even ones we don't need and now we have to carefully avoid those. Instead, could we strip out the irrelevant bindings from what `restore` gives us right at the very beginning? Something like:

```
diff --git a/src/dataflow/src/source/mod.rs b/src/dataflow/src/source/mod.rs
index 7e7a4b450..bf701bd07 100644
--- a/src/dataflow/src/source/mod.rs
+++ b/src/dataflow/src/source/mod.rs
@@ -1334,7 +1334,7 @@ where
             });

             let result = timestamp_histories.and_then(|timestamp_histories| {
-                let (mut valid_bindings, latest_bindings) = source_persist.restore().map_err(|e| {
+                let (mut valid_bindings, mut latest_bindings) = source_persist.restore().map_err(|e| {
                     SourceError::new(
                         sql_name.clone(),
                         SourceErrorDetails::Persistence(format!(
@@ -1344,6 +1344,22 @@ where
                     )
                 })?;

+                // Only work with the bindings this worker is responsible for.
+                valid_bindings.retain(|(source_ts, _)| {
+                    crate::source::responsible_for(
+                        &id.source_id,
+                        worker_id,
+                        worker_count,
+                        &source_ts.partition)
+                });
+
+                latest_bindings.retain(|(source_ts, _)| {
+                    crate::source::responsible_for(
+                        &id.source_id,
+                        worker_id,
+                        worker_count,
+                        &source_ts.partition)
+                });
                 let starting_offsets = source_persist.get_starting_offsets(valid_bindings.iter());

                 tracing::debug!(
@@ -1361,13 +1377,7 @@ where
                 );

                 for (pid, offset) in starting_offsets {
-                    if crate::source::responsible_for(
-                        &id.source_id,
-                        worker_id,
-                        worker_count,
-                        &pid) {
                         timestamp_histories.add_partition(pid, Some(offset));
-                    }
                 }

                 // We need to sort by offset and then timestamp because `add_binding()` will not allow
@@ -1375,14 +1385,6 @@ where
                 valid_bindings.sort_by(|a, b| (a.0.offset.offset, a.1).cmp(&(b.0.offset.offset, b.1)));

                 for (source_ts, assigned_ts) in valid_bindings.iter() {
-                    if !crate::source::responsible_for(
-                        &id.source_id,
-                        worker_id,
-                        worker_count,
-                        &source_ts.partition) {
-                        continue;
-                    }
-
                     // The timestamp bindings are potentially pre-seeded by bindings that we
                     // restored from the coordinator, if/when coordinator based timestamp
                     // persistence is active.
@@ -1535,7 +1537,6 @@ where
                     worker_id,
                     &mut timestamp_bindings_updater,
                     timestamp_histories,
-                    &consistency_info,
                     bindings_cap,
                     bindings_output,
                 );
@@ -1629,7 +1630,6 @@ where
                 worker_id,
                 &mut timestamp_bindings_updater,
                 timestamp_histories,
-                &consistency_info,
                 bindings_cap,
                 bindings_output,
             );
@@ -1883,7 +1883,6 @@ fn maybe_emit_timestamp_bindings(
     worker_id: usize,
     timestamp_bindings_updater: &mut Option<TimestampBindingUpdater>,
     timestamp_histories: &mut TimestampBindingRc,
-    consistency_info: &ConsistencyInfo,
     bindings_cap: &Capability<u64>,
     bindings_output: &mut OutputHandle<
         u64,
@@ -1901,19 +1900,14 @@ fn maybe_emit_timestamp_bindings(
     let changes = timestamp_bindings_updater.update(timestamp_histories);

     // Emit required changes downstream.
-    let to_emit = changes
-        .into_iter()
-        .filter(|((source_ts, _assigned_ts), _diff)| {
-            consistency_info.responsible_for(&source_ts.partition)
-        })
-        .map(|(binding, diff)| {
-            (
-                binding,
-                bindings_cap.time().clone(),
-                diff.try_into()
-                    .expect("could not convert i64 diff to isize"),
-            )
-        });
+    let to_emit = changes.into_iter().map(|(binding, diff)| {
+        (
+            binding,
+            bindings_cap.time().clone(),
+            diff.try_into()
+                .expect("could not convert i64 diff to isize"),
+        )
+    });

     // We're collecting into a Vec because we want to log and emit. This is a bit
     // wasteful but we don't expect large numbers of bindings.
```

Reply:
:ok_hand: